xds/internal/balancer/gracefulswitch/gracefulswitch_test.go

```diff
@@ -0,0 +1,1125 @@
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an UpdateState() call on the ClientConn")
+	case <-tcc.NewStateCh:
+	}
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an NewSubConn() call on the ClientConn")
+	case <-tcc.NewSubConnCh:
+	}
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an UpdateAddresses() call on the ClientConn")
+	case <-tcc.UpdateAddressesAddrsCh:
+	}
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an RemoveSubConn() call on the ClientConn")
+	case <-tcc.RemoveSubConnCh:
+	}
+	oldCurrent := gsb.balancerCurrent.Balancer.(*buildCallbackBal)
+
+	// Since the callback reports a state READY, this new inline balancer should
+	// be swapped to the current.
+	gsb.SwitchTo(buildCallbackBalancerBuilder{})
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an UpdateState() call on the ClientConn")
+	case <-tcc.NewStateCh:
+	}
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an NewSubConn() call on the ClientConn")
+	case <-tcc.NewSubConnCh:
+	}
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an UpdateAddresses() call on the ClientConn")
+	case <-tcc.UpdateAddressesAddrsCh:
+	}
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an RemoveSubConn() call on the ClientConn")
+	case <-tcc.RemoveSubConnCh:
+	}
+
+	// The current balancer should be closed as a result of the swap.
+	if err := oldCurrent.waitForClose(ctx); err != nil {
+		t.Fatalf("error waiting for balancer close: %v", err)
+	}
+
+	// The old balancer should be deprecated and any calls from it should be a no-op.
+	oldCurrent.newSubConn([]resolver.Address{}, balancer.NewSubConnOptions{})
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestShortTimeout)
+	defer cancel()
+	select {
+	case <-tcc.NewSubConnCh:
+		t.Fatal("Deprecated LB calling NewSubConn() should not forward up to the ClientConn")
+	case <-ctx.Done():
+	}
+}
+
+const balancerName1 = "mock_balancer_1"
+const balancerName2 = "mock_balancer_2"
+const verifyBalName = "verifyNoSubConnUpdateAfterCloseBalancer"
+const buildCallbackBalName = "callbackInBuildBalancer"
+
+type mockBalancerBuilder1 struct{}
+
+func (mockBalancerBuilder1) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	return &mockBalancer{
+		ccsCh:         testutils.NewChannel(),
+		scStateCh:     testutils.NewChannel(),
+		resolverErrCh: testutils.NewChannel(),
+		closeCh:       testutils.NewChannel(),
+		cc:            cc,
+	}
+}
+
+func (mockBalancerBuilder1) Name() string {
+	return balancerName1
+}
+
+type mockBalancerConfig struct {
+	serviceconfig.LoadBalancingConfig
+}
+
+// mockBalancer is a fake balancer used to verify different actions from
+// the gracefulswitch. It contains a bunch of channels to signal different events
+// to the test.
+type mockBalancer struct {
```

Comment:
Did you consider using the stub balancer defined in this [package](https://github.com/grpc/grpc-go/blob/master/internal/balancer/stub/stub.go#L102) instead of defining your own? I think using the stub balancer will significantly simplify things here.

Reply:
Doug and Menghan I think both agreed with this sentiment. I still don't get how using that simplifies things, especially since this has state and has other functions such as waitForXYZ. All the balancers clearly listed on this page with the methods they implement makes sense. I guess now you only have to define the methods that have logic. Let me know if you strongly prefer this, I like it as is.

Reply:
We used to (and still have some) mock balancers like this defined all over the place in tests. This is specifically the reason why the stub balancer package was created: so that each test can override just the methods that it wants and when you read one test you know exactly what is happening in there. I definitely feel that it helps with readability of the code. By simplify I meant that the code is easier to read. Now, to understand one test (or when I have to make a change to one test), I need to understand how the `mockBalancer` works. But I already know how the stub balancer works. Does that convince you?

Reply:
Rebuttal:
I'm not trying to get out of work here, but I've been switching it over to the stub balancer and I think for the use case of these unit tests I think declared balancers are much better than the stub balancer's API.

First, my unit tests are very heavy on testing which child balancer receives a forwarded update from grpc. This is possible because each created mock balancer has it's own channel, and you can call into balancerCurrent/Pending.waitForX() and it will wait on that specific channel. Current tests only have one channel which gets closed on/received (i.e. https://github.com/grpc/grpc-go/blob/master/internal/balancergroup/balancergroup_test.go#L517), and doesn't have this needed verification functionality of the correct balancer being forwarded to. In order to add this, I will need to declare two methods with sends on two specific channels, which adds to code bloat.

Also, having a single mock balancer with pretty easy functionality to learn (verifies any balancer.Balancer calls with appropriate waitFor calls (which as I mentioned before keeps it coupled per balancer, rather than closing on a channel we can't differentiate), and can ping upward to ClientConn), reused across all tests, feels cleaner than redefining for example
UpdateClientConnState by sending on a captured channel declared per test 10 separate times (again, multiplied by declaring the function twice per specific current/pending channel).

Also pinging upward from the balancer to Client Conn would require me to scale up the stub balancers API to also be a ClientConn, and to export the stub balancer. I need a concrete type to ping upward to the ClientConn and to be able to choose current or pending to ping to ClientConn, which isn't offered by the stub balancer.

I also find it cleaner to keep the verification instruments as state of the balancer
rather than declared per test and captured on (i.e. for closed https://github.com/grpc/grpc-go/pull/5207/files#diff-eba1c9a45b351799c36234d8a3d78d57b1f1f605bd2d8f0c40a9f065f86e41a3R991, keeping it as state vs.
declaring it in a test case).

Reply:
> First, my unit tests are very heavy on testing which child balancer receives a forwarded update from grpc.

Why is this not possible with the stub balancer API? Can't you register more than one stub balancer? Can you show a commit where you switch just one test from existing `mockBalancer` to the `stubBalancer` and maybe I can see what you are talking about.

Anyway, if you think this is becoming very painful, I'm OK with going with what we have and adding a TODO to say that we might want to use the stub balancer here instead.

Reply:
It is possible, but like I mentioned it repeats code many times (for both child balancers), and also requiring a method for every test that uses the stub balancer (since you mentioned in qazse an advantage of it was to do it per test case to keep methods scoped to test case).

Reply:
As far as tests are concerned, writing them in the most concise form is less important than making them readable.
go/tott/598
But up to you and you can totally leave a TODO if you dont want to do it now.